src/main.rs

```diff
@@ -34,52 +36,93 @@ header! { (XCachetToken, "X-Cachet-Token") => [String] }
 fn hook(req: HttpRequest) -> Box<Future<Item = HttpResponse, Error = Error>> {
     req.clone().json()                       // <- get JsonBody future
         .from_err()                          // <- automap to the error type we might want
-        .and_then(move |alert: AlertHook| {  // <- deserialized value
-            info!("{:?}", alert);
-
-            match alert.alerts.len() {
-                1 => info!("found one alert, as expected"),
-                _ => return Ok(HttpResponse::with_body(http::StatusCode::INTERNAL_SERVER_ERROR, "Expected exactly one alert."))
-
+        .and_then(move |alert_hook: AlertHook| {  // <- deserialized value
+            info!("{:?}", alert_hook);
+
+            let mut components : HashMap<i8, i8> = HashMap::new();
+            let mut responses = Vec::new();
+
+            for alert in alert_hook.alerts {
+                let current_severity : i8 = match components.get(&alert.annotations.component) {
+                    Some(severity) => *severity,
+                    None => 0 as i8,
+                }.clone();
+                let target_severity = match alert.status {
+                    Status::Firing => alert.annotations.severity,
+                    Status::Resolved => 1,
+                };
+                match current_severity {
+                    n if n < target_severity => {
+                        components.insert(alert.annotations.component, target_severity);
+                    }
+                    _ => {},
+                };
             }
 
-            //create map for json of cachet api call
-            let mut map = HashMap::new();
-            map.insert(
-                "status",
-                match alert.status {
-                    Status::Firing => alert.alerts[0].annotations.severity,
-                    Status::Resolved => 1,
-                },
-            );
-
-            //send http put to cachet
-            let client = reqwest::Client::new();
-            let response = match client.put(&format!(
-                "{endpoint}/api/v1/components/{component}",
-                endpoint = match std::env::var("CACHET_BASE_URL") {
-                    Ok(val) => val,
-                    Err(_) => String::from(include_str!("cachet_endpoint.txt")),
-                },
-                component = alert.alerts[0].annotations.component
-            ))
-                // read "Authorization: Bearer" token into "X-Cachet_Token"
-                .header(XCachetToken(match get_bearer_token(req) {
-                    Ok(token) => token,
+            let http_client = reqwest::Client::new();
+
+            for (component, status) in components {
```

Comment:
This is fine - while we can't group to nothing on component, we can group to just one component at a time, which should end up with a 1:1 mapping from (set of alert updates in a http request) -> (one component update).

But it's a good way of handling the case where we do mess up the config in "not a bad" way.

Possibly log the intermediate map as the useful debugging step in the middle - "{component 2 -> severity 4.  component 5 -> severity 1}" or some equivalent information.

Reply:
Like this?
```
[2018-07-23 16:45:44][DEBUG][prometheus_cachet_glue] Component statuses: {
    3: 1,
    4: 2,
    6: 1
}
```